core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -0,0 +1,271 @@
+                lengths[i] = length;
+                newByteCount += length;
+                offsets[positionCount + i + 1] = offsets[positionCount + i] + length;
+            }
+            hasNonNullValue = true;
+        }
+        copyBytes(block, lengths, positionArray, positions.size(), offsets, positionCount, newByteCount);
+    }
+
+    @Override
+    public void appendRle(RunLengthEncodedBlock block)
+    {
+        int rlePositionCount = block.getPositionCount();
+        if (rlePositionCount == 0) {
+            return;
+        }
+        int sourcePosition = 0;
+        ensurePositionCapacity(positionCount + rlePositionCount);
+        if (block.isNull(sourcePosition)) {
+            int offset = offsets[positionCount];
+            Arrays.fill(valueIsNull, positionCount, positionCount + rlePositionCount, true);
+            Arrays.fill(offsets, positionCount + 1, positionCount + rlePositionCount + 1, offset);
+            positionCount += rlePositionCount;
+
+            hasNullValue = true;
+            updateSize(rlePositionCount, 0);
+        }
+        else {
+            int startOffset = offsets[positionCount];
+            hasNonNullValue = true;
+            duplicateBytes(block.getValue(), sourcePosition, rlePositionCount, startOffset);
+        }
+    }
+
+    @Override
+    public Block build()
+    {
+        if (!hasNonNullValue) {
+            return new RunLengthEncodedBlock(NULL_VALUE_BLOCK, positionCount);
+        }
+        VariableWidthBlock result = new VariableWidthBlock(
+                positionCount,
+                Slices.wrappedBuffer(bytes, 0, currentOffset),
+                offsets,
+                hasNullValue ? Optional.of(valueIsNull) : Optional.empty());
+        reset();
+        return result;
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return retainedSizeInBytes;
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        return sizeInBytes;
+    }
+
+    private void copyBytes(Block block, int[] lengths, int[] positions, int count, int[] targetOffsets, int targetOffsetsIndex, int newByteCount)
+    {
+        ensureBytesCapacity(currentOffset + newByteCount);
+
+        for (int i = 0; i < count; i++) {
+            int position = positions[i];
+            if (!block.isNull(position)) {
+                int length = lengths[i];
+                Slice slice = block.getSlice(position, 0, length);
+                slice.getBytes(0, bytes, targetOffsets[targetOffsetsIndex + i], length);
+            }
+        }
+
+        positionCount += count;
+        currentOffset += newByteCount;
+        updateSize(count, newByteCount);
+    }
+
+    /**
+     * Copy {@code length} bytes from {@code block}, at position {@code position} to {@code count} consecutive positions in the {@link #bytes} array.
+     */
+    private void duplicateBytes(Block block, int position, int count, int startOffset)
```

Comment:
Do you need to pass startOffset since you also use current Offset below

Reply:
You could remove it altogether (I mean currentOffset). No need to track length twice (in offsets array and currentOffset)